kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KtTypes.kt

```diff
@@ -0,0 +1,239 @@
+import java.lang.reflect.Method
+import java.lang.reflect.Type
+
+internal val KmType.isNullable: Boolean get() = Flag.Type.IS_NULLABLE(flags)
+
+private fun defaultPrimitiveValue(type: Type): Any? =
+  if (type is Class<*> && type.isPrimitive) {
+    when (type) {
+      Boolean::class.java -> false
+      Char::class.java -> 0.toChar()
+      Byte::class.java -> 0.toByte()
+      Short::class.java -> 0.toShort()
+      Int::class.java -> 0
+      Float::class.java -> 0f
+      Long::class.java -> 0L
+      Double::class.java -> 0.0
+      Void.TYPE -> throw IllegalStateException("Parameter with void type is illegal")
+      else -> throw UnsupportedOperationException("Unknown primitive: $type")
+    }
+  } else null
+
+internal val KmType.canonicalName: String
+  get() {
+    return buildString {
+      val classifierString = when (val cl = classifier) {
+        is KmClassifier.Class -> createClassName(cl.name)
+        is TypeAlias -> createClassName(cl.name)
+        is TypeParameter -> arguments[cl.id].type?.canonicalName ?: "TypeVar(${cl.id})"
+      }
+      append(classifierString)
+
+      val args = arguments.joinToString(", ") {
+        "${it.variance?.name} ${it.type?.canonicalName ?: "*"}"
+      }
+
+      if (args.isNotBlank()) {
+        append('<')
+        append(args)
+        append('>')
+      }
+
+      // TODO not sure if we care about expressing the other type information here
+    }
+  }
+
+/**
+ * Creates a canonical class name as represented in Metadata's [kotlinx.metadata.ClassName], where
+ * package names in this name are separated by '/' and class names are separated by '.'.
+ *
+ * Example ClassName that we want to canonicalize: `"org/foo/bar/Baz.Nested"`.
+ *
+ * Local classes are prefixed with ".", but for Moshi's use case we don't deal with those.
+ */
+private fun createClassName(kotlinMetadataName: String): String {
+  require(!kotlinMetadataName.isLocal) {
+    "Local/anonymous classes are not supported!"
+  }
+  return kotlinMetadataName.replace("/", ".")
+}
+
+internal data class KtParameter(
+  val km: KmValueParameter,
+  val index: Int,
+  val rawType: Class<*>,
+  val annotations: List<Annotation>
+) {
+  val name get() = km.name
+  val declaresDefaultValue get() = Flag.ValueParameter.DECLARES_DEFAULT_VALUE(km.flags)
+  val isNullable get() = km.type!!.isNullable
+}
+
+internal data class KtConstructor(
+  val type: Class<*>,
+  val km: KmConstructor,
+  val jvm: Constructor<*>,
+  val parameters: List<KtParameter>,
+  val isDefault: Boolean
+) {
+  init {
+    jvm.isAccessible = true
+  }
+
+  fun <R> callBy(args: Map<KtParameter, Any?>): R {
+    val arguments = ArrayList<Any?>(parameters.size)
+    var mask = 0
+    val masks = ArrayList<Int>(1)
+    var index = 0
+
+    @Suppress("UseWithIndex")
+    for (parameter in parameters) {
+      if (index != 0 && index % Integer.SIZE == 0) {
+        masks.add(mask)
+        mask = 0
+      }
+
+      when {
+        args.containsKey(parameter) -> {
```

Comment:
any way you can save a map lookup? You’re doing containsKey and get, but you probably only need to look inside the map 1x not 2x. (worst case the map contains null, and you need to check containsKey() if get() returned null)

Reply:
264cdb5

Reply:
and 8f369d062c3fc0fc369fb677ce321d58681d5d1d (committed halfway)